src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Mvc3/AsyncBeginInvokeActionWrapper.cs

```diff
@@ -29,17 +30,30 @@ public AfterWrappedMethodDelegate BeforeWrappedMethod(InstrumentedMethodCall ins
             var controllerContext = instrumentedMethodCall.MethodCall.MethodArguments.ExtractNotNullAs<dynamic>(0);
             if (controllerContext != null)
             {
-                var controllerName = MvcRouteNamingHelper.TryGetControllerNameFromObject(controllerContext);
-                var actionName = MvcRouteNamingHelper.TryGetActionNameFromRouteParameters(instrumentedMethodCall.MethodCall, controllerContext.RouteData);
+                // IMPORTANT: Resist the urge to blindly refactor all of this code to use `var`
+                // IMPORTANT: We are being intentional with types over using `var` here due to
+                // IMPORTANT: the effects of handling a `dynamic` object
+                string controllerName = MvcRouteNamingHelper.TryGetControllerNameFromObject(controllerContext);
+                string actionName = MvcRouteNamingHelper.TryGetActionNameFromRouteParameters(instrumentedMethodCall.MethodCall, controllerContext.RouteData);
 
                 var httpContext = controllerContext.HttpContext;
                 if (httpContext == null)
+                {
                     throw new NullReferenceException("httpContext");
+                }
 
-                var transactionName = string.Format("{0}/{1}", controllerName, actionName);
+                string transactionName = controllerName + "/" + actionName;
                 transaction.SetWebTransactionName(WebTransactionType.MVC, transactionName, TransactionNamePriority.FrameworkLow);
 
-                var segment = transaction.StartMethodSegment(instrumentedMethodCall.MethodCall, controllerName, actionName);
+                ISegment segment = transaction.StartMethodSegment(instrumentedMethodCall.MethodCall, controllerName, actionName);
+
+                string fullControllerName = MvcRouteNamingHelper.TryGetControllerFullNameFromObject(controllerContext);
```

Comment:
Minor but by duplicating the `TryGet...` pattern for the full name we are executing some of the same code to get the type where we could try to get the type once and then use the short and long names. Perhaps changing existing usages would be painful, though. I do believe `GetType()` has some magic that makes it pretty fast (and only allocate once if I remember right, the bigger concern) and this should not be a hot path.